File: wake1/lib/wake/modules/copy.py
import shutil
import wake.util as util

def name():
    return "copy"

def wants(filename):
    # The copy module is a catch-all:
    return True

def produces(filename):
    return [util.src2out(filename)]

def process(filename):
    outfile = produces(filename)[0]
    util.check_dir(outfile)
    shutil.copy(filename, outfile)
    print("cp " + filename + " -> " + outfile)

File: src/main/java/wake/plugins/Copy.java
package wake.plugins;

import java.nio.file.Files;
import java.nio.file.StandardCopyOption;
import java.util.ArrayList;
import java.util.List;

import wake.core.Plugin;
import wake.core.WakeFile;

public class Copy implements Plugin {

    /* File copy has no dependencies: */
    private List<WakeFile> dependencies = new ArrayList<>();

    public Copy() {

    }

    @Override
    public String name() {
        return "Copy";
    }

    @Override
    public boolean wants(WakeFile file) {
        /* We'll copy anything we can get our hands on! */
        return true;
    }

    @Override
    public List<WakeFile> requires(WakeFile file) {
        return this.dependencies;
    }

    @Override
    public List<WakeFile> produces(WakeFile file) {
        WakeFile output = file.toOutputFile();
        ArrayList<WakeFile> outputs = new ArrayList<>();
        outputs.add(output);
        return outputs;
    }

    @Override
    public List<WakeFile> process(WakeFile file) throws Exception {
        WakeFile output = file.toOutputFile();

        Files.copy(file.toPath(), output.toPath(),
                StandardCopyOption.REPLACE_EXISTING);

        List<WakeFile> outputs = new ArrayList<>();
        outputs.add(output);
        return outputs;
    }
}
